[tests] EndToEnd - add test scenarios

radical opened the pull request:
Test scenarios allow defining different groupings for tests to be run. This is useful for CI to run the tests in separate helix items.

```csharp
        TestResourceNames resourcesToInclude = TestScenario switch
        {
            "oracle" => TestResourceNames.oracledatabase,
            "cosmos" => TestResourceNames.cosmos,
            "scenario0" => TestResourceNames.kafka
                              | TestResourceNames.mongodb
                              | TestResourceNames.rabbitmq
                              | TestResourceNames.redis
                              | TestResourceNames.postgres
                              | TestResourceNames.efnpgsql
                              | TestResourceNames.mysql
                              | TestResourceNames.efmysql
                              | TestResourceNames.sqlserver,
            "" or null => TestResourceNames.All,
            _ => throw new ArgumentException($"Unknown test scenario '{TestScenario}'")
        };
```

Once `eforacle` is added it can run with the `oracle` scenario.

The default is to run all the tests.

Also:
- Enable `cosmos` E2E tests on helix
- make `TestResourceNames` into a `[Flag]` enum so multiple can be expressed at the same time.
- And support dumping docker container logs for resources where the resource name does not match the container name prefix. For example `efnpgsql` vs `postgres`.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3113)

radical committed:
[tests] Make E2E tests a special case of workload tests

This makes EndToEnd tests a special case of workload tests, IOW, tests
that need to use a sdk+workload-installed-from-artifacts. The reason
being that these tests need some special logic to queue up multiple
helix items to run multiple test scenarios, which differ only in their
setup and not the test assembly itself.

In follow up PRs with other workload test projects being added, they
will default to adding one item but with the sdk+workload.

radical committed:
[tests] Add test scenarios to EndToEnd tests

These tests start up multiple containers and those can be too heavy for
the helix machines. Instead, adding test scenarios allows running the
tests on multiple helix machines with specific containers.

For example:
- scenario0 - this is the set of the "light" containers
- cosmos - this one takes a lot of time to get ready, thus runs by
  itself

Also, this renames `pomelo` service in the tests to `efmysql`.

radical committed:
[tests] Fix retries in the testproject services

In the following code snippet, the old code "retries 60 times", which
can total up to 2 mins. And 2 mins is the timeout used by the actual
EndtoEnd tests invoking these services.

```
+++ b/tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs
@@ -18,7 +18,7 @@ private static async Task<IResult> VerifyCosmosAsync(CosmosClient cosmosClient)
             var policy = Policy
                 .Handle<HttpRequestException>()
                 // retry 60 times with a 1 second delay between retries
-                .WaitAndRetryAsync(60, retryAttempt => TimeSpan.FromSeconds(1));
+                .WaitAndRetryAsync(20, retryAttempt => TimeSpan.FromSeconds(1));

             var db = await policy.ExecuteAsync(
                 async () => (await cosmosClient.CreateDatabaseIfNotExistsAsync("db")).Database);
```

Instead, reduce the number of attempts to a more reasonable number.

radical commented on tests/testproject/TestProject.IntegrationServiceA/SqlServer/SqlServerExtensions.cs:
```diff
@@ -18,7 +18,7 @@ private static async Task<IResult> VerifySqlServerAsync(SqlConnection connection
                 .Handle<SqlException>()
                 // retry 60 times with a 1 second delay between retries
-                .WaitAndRetryAsync(60, retryAttempt => TimeSpan.FromSeconds(1));
+                .WaitAndRetryAsync(10, retryAttempt => TimeSpan.FromSeconds(1));
```
@eerhardt @sebastienros Is there a better way to express a "overall timeout" for the attempts here, so it all completes under n mins? Otherwise, this keeps retrying and the caller (test) fails because of it's own timeout (2mins).

radical committed:
cleanup

radical committed:
Merge branch 'dotnet:main' into e2e-scenarios-staging

mitchdenny committed:
Remove obsolete APIs. (#3329)

djonser committed:
In AWS Playground AppHost, add QueuePolicy to ChatMessagesQueue so that SNS Subscription will work. (#3333)

eerhardt committed:
Remove Azure SDK feed from nuget.config (#3327)

radical committed:
Cleanup

eerhardt committed:
Fix ConfigurationSchemaGenerator to use correct TimeSpan format (#3320)

* Fix ConfigurationSchemaGenerator to use correct TimeSpan format

'duration' isn't an acceptable description of how to represent a TimeSpan. Instead use a custom regular expression to represent TimeSpan formats.

* Add tests for the timespan regex.

* Fix outdated JSON config schema tests

* Fix more tests

eerhardt committed:
Update AspNetCore.HealthChecks.* to latest versions (#3338)

The new versions have specific builds for the .NETCoreApp TFM, which fixes some dependency issues like not having the 8.0.0 version of Microsoft.Bcl.AsyncInterfaces. See https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/issues/2180.

dotnet-bot committed:
Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2420391 (#3335)

* Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2419838

* Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2419838

* Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2419838

* Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2419838

* Localized file check-in by OneLocBuild Task: Build definition ID 1309: Build ID 2419960

eerhardt committed:
Dashboard should request log streams only when the user gestures to display them (#3235)

Change the AppHost to only subscribe to DCP's logs when there is a subscriber for the logs.

Fix #2789

* Show logs

* Remove the "StreamingLogger" because that causes subsequent subscribers to not see what was already written.

Fix duplicate logs issue by clearing out the backlog when the last subscriber leaves.

* Fix a concurrency issue with the backlog. Ensure the backlog is only snap shotted after subscribing to the log.

Fix existing tests for new functionality and add an additional test.

* PR feedback

Only clear the backlog on containers and executables.

* PR feedback.

Move lock out of async iterator.

* Clean up code.

- Remove count and instead check if the delegate is null to indicate whether there are subscribers or not.
- Remove the separate IAsyncEnumerable classes and just use `async IAsyncEnumerable` methods.

* Fix a race at startup when logs aren't available.

Employ a 2nd loop that listens for both "has subscribers" and "logs available".

* Simplify ResourceNotificationService.WatchAsync.

* Fix test build

* Address PR feedback

- Set SingleReader=true on the logInformationChannel.
- Add comment and assert that LogsAvailable can only turn true and can't go back to false.

---------

Co-authored-by: David Fowler <davidfowl@gmail.com>

RussKie committed:
Publish build assets in the build stage (#3334)

karolz-ms committed:
Improve service address allocation (#3294)

* Improve service address allocation
Should fix https://github.com/dotnet/aspire/issues/3265

davidfowl committed:
Added ability to resolve target port (#3305)

* Added ability to resolve target port
- Change how we process and expose target ports. This change makes it possible to avoid using any internals of DCP to access and resolve target ports from any resource.
- More launch profile processing to the project resource defaults.
- Remove logic in the manifest writing that wrote a special env variable. This should just work now  like everything else.
- Exposed TargetPort and Exists on EndpointReference.
- Use this new target property with dapr.
- Added a ReferenceExpressionBuilder which makes it possible to dynamically build a ReferenceExpression.
- Endpoints and env are evaulated earlier now so change remove
tests that add default launch profile on real projects
without allocating those endpoints.

* Remove more uses of TestProgram

* Fix tests

JamesNK committed:
Add token auth to dashboard frontend (#3213)

mitchdenny committed:
Tweak Bicep string name. (#3331)

* Tweak Bicep string name.

* Fix up test.

JamesNK committed:
Improve writing dashboard startup config failure messages (#3243)

JamesNK committed:
Update OpenTelemetry packages to 1.8.0 (#3351)

* Update OpenTelemetry packages to 1.8.0

* Update template versions

eerhardt commented on Directory.Build.props:
```diff
@@ -45,8 +45,9 @@
     <DcpDir>$(NuGetPackageRoot)microsoft.developercontrolplane.$(BuildOs)-$(BuildArch)/$(MicrosoftDeveloperControlPlanedarwinamd64PackageVersion)/tools/</DcpDir>
 
-    <TestArchiveTestsDir>$([MSBuild]::NormalizeDirectory($(ArtifactsDir), 'helix', 'tests'))</TestArchiveTestsDir>
+    <TestArchiveTestsDir Condition="'$(TestArchiveTestsDir)' == ''">$([MSBuild]::NormalizeDirectory($(ArtifactsDir), 'helix', 'tests'))</TestArchiveTestsDir>
```
Can these "Test*" properties be moved to the tests\Directory.Build.props|targets? That way they don't clutter up the root.

The root props and targets should only be for things that apply across all projects in the repo.

eerhardt commented on Directory.Build.targets:
```diff
@@ -5,7 +5,7 @@
     <PackageReadmeFile Condition="'$(PackageReadmeFile)' == '' And '$(ReadMeExists)' == 'true'">README.md</PackageReadmeFile>
 
-    <TestArchiveTestsDir Condition="'$(IsWorkloadTestProject)' == 'true'">$(TestArchiveTestsDirForWorkloadTests)</TestArchiveTestsDir>
+    <TestArchiveTestsDir Condition="'$(TestArchiveTestsDir)' == '' and '$(IsWorkloadTestProject)' == 'true'">$(TestArchiveTestsDirForWorkloadTests)</TestArchiveTestsDir>
```
This is weird that this property is set in both the `.props` and the `.targets`. Why both?

Can this condition ever be true since it is already defaulted in the `.props`?

eerhardt commented on tests/send-to-helix-workload-tests.proj:
```diff
@@ -51,18 +64,23 @@
 
         <!-- Download results file so coverage files can be extracted -->
-        <DownloadFilesFromResults>logs/%(FileName).cobertura.xml</DownloadFilesFromResults>
+        <!-- FIXME: does sharing the TEST_NAME affect the reports? Yes, TestNameEnvVar - gets used for these reports. Make sure they end up in different directories! -->
```
What's the plan to address this?

radical replied:
this can be removed now as I had checked the code-coverage reports and they were unaffected.

eerhardt commented on tests/send-to-helix-workload-tests.proj:
```diff
@@ -51,18 +64,23 @@
            Text="Could not find dotnet-coverage tool. %24(_DotNetCoverageToolPath)=$(_DotNetCoverageToolPath)" />
 
     <ItemGroup>
+      <_E2ETestScenarios Include="scenario0" />
```
Can we give this a better name than `scenario0`?

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:
```diff
@@ -42,30 +43,40 @@ public Task VerifyComponentWorks(TestResourceNames resourceName)
-    [InlineData(TestResourceNames.oracledatabase)]
-    public Task VerifyComponentWorksDisabledOnCI(TestResourceNames resourceName)
+    [Theory]
+    [Trait("scenario", "sqlserver")]
```
Why are we splitting SqlServer out?

eerhardt commented on tests/testproject/TestProject.IntegrationServiceA/Cosmos/CosmosExtensions.cs:
```diff
@@ -18,7 +18,7 @@ private static async Task<IResult> VerifyCosmosAsync(CosmosClient cosmosClient)
                 .Handle<HttpRequestException>()
                 // retry 60 times with a 1 second delay between retries
-                .WaitAndRetryAsync(60, retryAttempt => TimeSpan.FromSeconds(1));
+                .WaitAndRetryAsync(20, retryAttempt => TimeSpan.FromSeconds(1));
```
The comment is out of date.

Also, why are we lowering this retry time?

eerhardt commented on tests/send-to-helix-workload-tests.proj:
```diff
@@ -2,6 +2,9 @@
 
```
Is this file really a `.proj` file? Since it is imported by the `send-to-helix.proj`, isn't it more of a `.targets` file?

adamint committed:
Make sure copy button is tab-accessible and visible on tab (#3192)

Co-authored-by: Adam Ratzman <adamratzman@microsoft.com>

radical committed:
rename send-to-helix-workload-tests.proj to .targets

radical committed:
rename scenario0 to basicservices, and merge sqlserver with basicservices

radical committed:
Move $(Test*) to test/Directory.Build.props

.. and explicitly set the `$(TestArchiveTestsDir)` to
`$(TestArchiveTestsDirForEndToEndTests)` in
`EndToEnd.Tests.csproj`.

radical committed:
Revert timeout changes from services. this will get addressed in a follow up PR

radical committed:
remove resolved fixme

radical commented:
Hopefully, I didn't break the tests!

radical committed:
Merge remote-tracking branch 'origin/main' into e2e-scenarios-staging

radical committed:
fix scenario0 everywhere

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -80,9 +82,13 @@ public async Task InitializeAsync()
         string processArguments = $"run --no-build -- ";
-        if (GetResourcesToSkip() is var resourcesToSkip && resourcesToSkip.Count > 0)
+        _resourcesToSkip = GetResourcesToSkip();
+        if (_resourcesToSkip > 0)
```
```suggestion
        if (_resourcesToSkip != TestResourceNames.None)
```

eerhardt commented on tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:
```diff
@@ -316,23 +336,57 @@ public void EnsureAppHostRunning()
+            {
+                resourcesToSkip |= TestResourceNames.cosmos;
+                resourcesToSkip |= TestResourceNames.oracledatabase;
+                resourcesToSkip |= TestResourceNames.sqlserver;
```
```suggestion
```

radical replied:
good catch!

eerhardt commented on tests/send-to-helix-workload-tests.targets:
```diff
@@ -51,18 +64,21 @@
 
     <ItemGroup>
+      <_E2ETestScenarios Include="basicservices" />
+      <_E2ETestScenarios Include="cosmos" />
```
? I thought we didn't run this in CI.

or is the intention to start running it now?

eerhardt commented on tests/send-to-helix-workload-tests.targets:
```diff
@@ -51,18 +64,21 @@
 
         <!-- Download results file so coverage files can be extracted -->
-        <DownloadFilesFromResults>logs/%(FileName).cobertura.xml</DownloadFilesFromResults>
+        <DownloadFilesFromResults>logs/Aspire.EndToEnd.Tests.cobertura.xml</DownloadFilesFromResults>
```
Isn't this name going to conflict between multiple _E2ETestScenarios ?

radical replied:
hmph yeah:
```
Copying /mnt/vss/_work/1/s/artifacts/helixresults/cc085e87-12fa-4085-8946-dc76e4f0f4c1/Aspire.EndToEnd.Tests-cosmos/logs/Aspire.EndToEnd.Tests.cobertura.xml to /mnt/vss/_work/1/s/artifacts/CodeCoverage/Aspire.EndToEnd.Tests.cobertura.xml
File /mnt/vss/_work/1/s/artifacts/helixresults/cc085e87-12fa-4085-8946-dc76e4f0f4c1/Aspire.EndToEnd.Tests-scenario0/logs/Aspire.EndToEnd.Tests.cobertura.xml already exist at /mnt/vss/_work/1/s/artifacts/CodeCoverage/Aspire.EndToEnd.Tests.cobertura.xml
File /mnt/vss/_work/1/s/artifacts/helixresults/cc085e87-12fa-4085-8946-dc76e4f0f4c1/Aspire.EndToEnd.Tests-sqlserver/logs/Aspire.EndToEnd.Tests.cobertura.xml already exist at /mnt/vss/_work/1/s/artifacts/CodeCoverage/Aspire.EndToEnd.Tests.cobertura.xml
```

radical replied:
fixed.

eerhardt reviewed (approved):
Just a couple minor suggestions.

radical committed:
Update tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Update tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Use scenario name in cobertura.xml for e2e tests to differentiate between them

radical committed:
fix reference to renamed property

joperezr commented:
Thanks for this @radical . I'm assuming we want to backport it for P6 so we get coverage in release branches too?

joperezr commented:
/backport to release/8.0
